cirq-core/cirq/protocols/json_serialization.py

```diff
@@ -386,6 +387,23 @@ def default(self, o):
                 'index': o.index,
             }
 
+        # datetime
+        if isinstance(o, datetime.datetime):
+            if o.tzinfo is not None:
+                # coverage: ignore
+                raise TypeError("datetime tzinfo is not serializable")
+            return {
+                'cirq_type': 'datetime.datetime',
+                'year': o.year,
+                'month': o.month,
+                'day': o.day,
+                'hour': o.hour,
+                'minute': o.minute,
+                'second': o.second,
+                'microsecond': o.microsecond,
+                'fold': o.fold,
+            }
```

Comment:
I would suggest serializing the timestamp instead:
```suggestion
        if isinstance(o, datetime.datetime):
            if o.tzinfo is not None:
                # coverage: ignore
                raise TypeError("datetime tzinfo is not serializable")
            return {
                'cirq_type': 'datetime.datetime',
                'timestamp': o.timestamp(),
            }
```
For deserialization, you'll need to call `fromtimestamp` instead of the class constructor:
```
datetime.datetime.fromtimestamp(json['timestamp'])
```
Note that without tzinfo, serializing datestamps with calendar values like year, month, day means the result will be interpreted in the locale timezone at both serialization and deserialization time, which may be different. The timestamp, on the other hand, is based on the unix epoch, so this will ensure that the deserialized time is the same instant, even though it might have different calendar entries after deserialization if the locale timezone is different. Also, with timestamp you don't have to assert that `o.tzinfo` is None because both tz-aware and "naive" datetimes have a timestamp.

Reply:
I'll split this out. I think it's really important to be able to serialize timestamps and it's crazy we're missing this functionality.

Having looked a little closer: what do you think about _only_ allowing the serialization of UTC datetimes? when tzinfo=None, there's a ton of ambiguity and it seems harder to accurately round-trip datetimes via unix timestamp, which converts to utc

Reply:
```python
    naive_ts = datetime.datetime.now()
    utc_ts = naive_ts.astimezone(datetime.timezone.utc)
    assert naive_ts != utc_ts, 'tzinfo'
    assert naive_ts.timestamp() == utc_ts.timestamp()
    assert (utc_ts.day, utc_ts.hour) > (naive_ts.day, naive_ts.hour), 'at least in PDT'

    re_utc = datetime.datetime.fromtimestamp(utc_ts.timestamp())
    re_naive = datetime.datetime.fromtimestamp(naive_ts.timestamp())

    assert re_utc == re_naive, 'roundtripping turns to utc'
    assert re_utc != utc_ts, 'roundtripping loses tzinfo'
    assert naive_ts == re_naive, 'works, but only because my local time isnt changing(?)'
```

Reply:
#5274